Generalizing GetGradient, GetCurl, and GetDivergence [gf-getgradient-dev]

mlstowell opened the pull request:
This is a follow-on PR to #1194 which uses the same technique to generalize the `GridFunction::Get[Gradient|Curl|Divergence]` methods so that they also work on boundary elements.
<!--GHEX{"id":1495,"author":"mlstowell","editor":"v-dobrev","reviewers":["psocratis","vladotomov"],"assignment":"2020-05-27T15:42:08-07:00","approval":"2020-06-09T01:40:02.188Z","merge":"2020-06-16T01:40:02.188Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1495](https://github.com/mfem/mfem/pull/1495) | @mlstowell | @v-dobrev | @psocratis + @vladotomov | 05/27/20 | 06/08/20 | ⌛due 06/15/20 | |
<!--ELBATXEHG-->

psocratis commented on fem/coefficient.cpp:
```diff
@@ -212,15 +212,16 @@ void GradientGridFunctionCoefficient::Eval(
    const GridFunction *gf)
    : VectorCoefficient ((gf) ?
-                        gf -> FESpace() -> GetMesh() -> SpaceDimension() : 0)
+                        (2 * gf -> FESpace() -> GetMesh() -> SpaceDimension()
```
So just to confirm if I understand this correctly... in case of a 2D problem, a VectorCoefficient of dim=1 would be used instead of a Coefficient...is that right?

mlstowell replied:
That is correct.

najlkin replied:
What will happen in 1D? The vector dimension will be negative, right? Maybe add an assert or use the logic `dim == 3 ? 3 : 1`, which is applied elsewhere and the object will be valid and it will just die at the evaluation.

mlstowell replied:
@najlkin , that's a good point.  The curl will always be zero in 1D and it's only defined in 2D or 3D.  I added an `MFEM_VERIFY` to check the space dimension and issue an error if the dimension is not 2 or 3.

najlkin commented on fem/gridfunc.cpp:
```diff
@@ -1338,107 +1338,253 @@ void GridFunction::GetVectorGradientHat(
+      break;
+      case ElementTransformation::BDR_ELEMENT:
+      {
+         // In order to capture the normal component of the gradient we
```
Is it only the normal component? According to the code, it does not look so.

mlstowell replied:
No, it's not just the normal component.  Perhaps the comment is unclear.  I'll expand it.

najlkin commented on fem/gridfunc.cpp:
```diff
@@ -1338,107 +1338,253 @@ void GridFunction::GetVectorGradientHat(
+
+         // Evaluate in neighboring element
+         ElementTransformation & T1 = FET->GetElement1Transformation();
+         return GetDivergence(T1);
```
The same as above, it is not reduced by Stokes's theorem to the lower dimension.

mlstowell replied:
No, we compute the divergence of every basis function in the element and sum the results together to compute the total divergence at the desired point.

mlstowell commented on fem/gridfunc.cpp:
```diff
@@ -1338,107 +1338,253 @@ void GridFunction::GetVectorGradientHat(
+         FET->SetIntPoint(&fip);
+         ElementTransformation & T1 = FET->GetElement1Transformation();
+
+         return GetDivergence(T1);
```
This returns the full divergence.  The derivative in the normal direction depends on all of the element's DoFs not just those on the face where as the tangential derivatives could be calculated using only face DoFs.  These comments were only meant to highlight why we go the extra trouble of evaluating these derivatives in the neighboring element.  They should have been more clear.  Thanks!

najlkin replied:
But actually that is maybe what you want calculate, because it is also mathematically incorrect to evaluate it inside. For a derivative, you need a neighborhood of the point, which you do not have at the boundary of an element. This is especially counterintuitive for discontinuous functions, where you solve it by taking the derivative only on the first side, but why? it could be the other one, or average of both, etc.
To summarize, I am not sure that the generalization is that straightforward. Basically, there are two approaches to it, which I see:
- reduction to the lower dimension of the boundary/face element through Stokes' theorem. In the case of the divergence, it would the normal component for the boundary elements and the normal jump for internal faces. This is what came to my mind at first when reading about the generalization.
- evaluation of the derivative on the inner volume, but the user should have freedom in choice of the behavior or it should be somehow prevalent option. However, I do not see how the inner volume divergence on the first side of the face is useful practically.

mlstowell replied:
This method is only implemented for element interiors and the mesh boundary so there is no other side of the face to be concerned with.

I agree that computing the divergence on the boundary is odd and may not be useful in most cases.  I would prefer to keep this method as it is but offer two other alternatives that calculate either `du/dn` or `\nabla_t u`.

najlkin replied:
I do not know, what do others think? @vladotomov , @psocratis , @jakubcerveny , should be the generalization of the differential operators just their mere evaluation on the inner volume of the element OR reduction to the lower dimension by means of the Stokes' theorem, i.e. normal jump for divergence and tangential jump for curl? I think the latter might be useful for someone and can be implemented also for the inner faces. This is something different than the generalization of GetValue() in #1194 , where point-wise evaluation at the boundary is the only meaningful thing you may do (except DGs, but the trace is what you normally mean).

vladotomov replied:
@najlkin I think that should probably be a different PR with different function names.

vladotomov commented on tests/unit/fem/test_get_value.cpp:
```diff
@@ -1829,4 +1904,1669 @@ TEST_CASE("3D GetVectorValue",
+          "[DivergenceGridFunctionCoefficient]")
+{
+   int log = 1;
+   int n = 1;
```
I noticed most tests use a 1-element mesh. Does it make sense to test on at least 2 elements, and maybe one of them locally refined?

mlstowell replied:
Honestly, I was a little hesitant to use such a small mesh.  However, my justification is two-fold: I want to keep the unit tests as quick as possible and the only thing that is likely to break on a larger mesh would be the element transformation which is tested elsewhere.

I can certainly increase the mesh sizes if you like.  After all, the tests are meant to find errors that we don't predict.

tzanio commented:
Merged in `next` for testing...

tzanio commented:
LGTM

v-dobrev commented on fem/gridfunc.cpp:
```diff
@@ -1467,15 +1616,63 @@ void GridFunction::GetGradients(ElementTransformation &tr,
+      {
+         // In order to capture the normal component of the gradient we
+         // must evaluate it in the neighboring element.
+         FaceElementTransformations * FET =
+            fes->GetMesh()->GetBdrFaceTransformations(T.ElementNo);
+
+         // Boundary elements and Boundary Faces may have different
+         // orientations so adjust the integration point if necessary.
+         int o = 0;
+         if (fes->GetMesh()->Dimension() == 3)
+         {
+            int f;
+            fes->GetMesh()->GetBdrElementFace(T.ElementNo, &f, &o);
+         }
+
+         IntegrationPoint fip;
+         be_to_bfe(FET->GetGeometryType(), o, T.GetIntPoint(), fip);
```
There is another way to transform boundary quadrature points to volume quadrature points that is used in other places. The two `Mesh` methods that can be used are: `GetBdrElementAdjacentElement` and `GetLocalFaceTransformation`, see e.g. https://github.com/mfem/mfem/blob/b8ab8fce47d9081a3a7bf311b1025f64af9ce54d/mesh/mesh.cpp#L454-L459

We should probably unify these at some point.

v-dobrev commented on fem/gridfunc.cpp:
```diff
@@ -1467,15 +1616,63 @@ void GridFunction::GetGradients(ElementTransformation &tr,
+         be_to_bfe(FET->GetGeometryType(), o, T.GetIntPoint(), fip);
+
+         FET->SetIntPoint(&fip);
+         ElementTransformation & T1 = FET->GetElement1Transformation();
+
+         GetVectorGradient(T1, grad);
```
The integration point is never transferred from the face (i.e. from `fip`) to the volume. This issue seems to also affect the other methods that are extended in this branch.

mlstowell replied:
The new implementation of `FaceElementTransformations::SetIntPoint` transforms and sets the integration points in the volume transformations as well.

v-dobrev replied:
Okay, I didn't realize there's now a new version of `SetIntPoint` in `FaceElementTransformations`. This is then not an issue.

v-dobrev replied:
One issue with this new version of `FaceElementTransformations::SetIntPoint` is that now most of the DG integrators will repeat these transformations -- they do these transformations explicitly, if needed.

v-dobrev replied:
One simple solution is to rename the new `FaceElementTransformations::SetIntPoint` to something like `FaceElementTransformations::SetAllIntPoints`. That way the DG integrators will not need to change and it will be more clear that we set all (face+element) quadrature point with the new name. What do you think @mlstowell?

mlstowell replied:
The DG integrators should not need to change.  They access `FET->Face->SetIntPoint` which does not compute and set the volume integration points.  I wouldn't mind adding a boolean argument to `FaceElementTransformations::SetIntPoint(const IntegrationPoint *ip, bool transform_all = true)` to disable this feature.

v-dobrev replied:
You are right, in some places the deprecated `Face` member is used -- these should probably be updated. However, in many places `FaceElementTransformations::SetIntPoint` is used -- I see 5 places in `bilininteg.cpp`, 4 in `lininteg.cpp` and a few more.

Adding a default parameter does not help when reading the code -- when I see a call to `SetIntPoint` with one parameter I'll have to remember to check the type of the object to understand what is actually being called.

mlstowell replied:
You make a valid point, It will take some getting used to.  From my point of view this is a natural consequence of promoting `FaceElementTransformations` from a struct to a class.  When it was a struct it was natural to configure the various pieces separately because they were separate objects which just happened to be passed around as a struct for convenience.  Now that they are parts of a single object it makes sense, at least to me, to configure the object as a whole by default.  Adding more fine-grained control for experts, as we do with the configuration mask, is the right choice, I think.  But it does require more care.

However, I agree that readability is very important.  How about this:
```
/// Sets the face integration point
void SetFaceIntPoint(const IntegrationPoint *ip);

/// Sets the integration point in element 1
void SetElement1IntPoint(const IntegrationPoint *ip);

/// Sets the integration point in element 2
void SetElement2IntPoint(const IntegrationPoint *ip);

/// Set the integration point in the face and elements 1 and 2 if present
void SetIntPoint(const IntegrationPoint *ip);
```

v-dobrev replied:
I don't mind adding the first three methods, even though they may not be needed at the moment.

However, for the last method, I think it is best to use a different name and keep `SetIntPoint` as inherited from the base class `ElementTransformation`.

mlstowell replied:
I'm sorry, @v-dobrev, but I have to disagree.  We must think of the consequences of using these methods incorrectly.  In the current implementation if a developer calls `SetIntPoint` when they don't actually need the element transforms configured then they take a performance hit.  This is not ideal.  If we only set the point in the `Face` object but the developer expects the entire object to be usable there are two possible outcomes.  First, the `ElementTransformation` might crash when it accesses the NULL `IntegrationPoint` pointer.  This is ok but not ideal.  The more likely outcome is that the `ElementTransformation` has been used before and has a non-NULL `IntegrationPoint` pointer.  In this case the `ElementTransformation` will quietly return the wrong results. This type of error is a nightmare to track down if it is even noticed (and that's a BIG "if").

v-dobrev replied:
@mlstowell, by using the same name, you are trying to change the meaning of the method `SetIntPoint` in a derived class and that function is (intentionally, by design) not virtual. Derived classes are not supposed to do that because it creates confusion, as I mentioned above.

If you want to have a "safe" method, you can add a new virtual method (with a different name) in the base class and override that in derived classes. This is also a "safer" solution because it will work correctly even if you call the method by a reference/pointer to the base class.

In any case, I think we your input, @psocratis and @vladotomov, to help resolve this disagreement.

mlstowell replied:
What is the reason for `SetIntPoint` not being virtual?

v-dobrev replied:
@mlstowell, I can give you my reasons but at this point it is not going to lead us anywhere.

I think we should wait for the reviewers' (@psocratis and @vladotomov) input.

mlstowell replied:
That's fine.  I'm not trying to be argumentative.  I was just curious.

psocratis replied:
I am kind of torn between the two solutions... 
May I ask the following?

1) How bad is the performance hit in the case of the current implementation?
2) Is there a significant performance hit under what @mlstowell suggests for DG integrators? 
3) Finally, is there a real downside of using a new FaceElementTransformations::(Some different name other than SetIntPoint) ? 

If the answer to (3) is no, I am leaning towards that. 

Socratis

mlstowell commented:
1. The performance hit I referred to involves computing the `IntegrationPoint` in the neighboring volume elements from the `IntegrationPoint` on the face.  This means we evaluate the lowest order scalar basis functions on the face and compute a small matvec to get the higher dimensional integration point.  This can only be considered a "hit" when we don't need this information.

2. As @v-dobrev pointed out, this `FaceElementTransformations:: SetIntPoint` method is used in 5 DG `BilinearformIntegrators`.  In 4 of these integrators this transformation is required so no performance hit is incurred.  In the 5-th integrator, the `BoundaryMassIntegrator`, the neighboring `ElementTransformation` objects are not needed explicitly so it might appear that this transformation is unnecessary.  However, the transformation is passed to a coefficient which might need to be evaluated in the neighboring element.  So, in this one case there is a potential performance hit which we pay to make sure the user's coefficient has all the information it might need.

3. The downside is that users might implement their own integrators and use `FaceElementTransformations::SetIntPoint` expecting it to set the volume integration points as well.  When it doesn't their computations in neighboring elements will be wrong and they won't know that.

v-dobrev commented:
@psocratis,

Let me try to summarize what my concerns and suggestions are:

For `BoundaryMassIntegrator`, my suggestion is to avoid always transforming the face point to the volume in the call to `Trans.SetIntPoint(&ip);` here:
https://github.com/mfem/mfem/blob/bf62d2923cc7390c0c78035dcebf3a6789187264/fem/bilininteg.cpp#L928-L933

Instead that transformation should happen only when needed, e.g. here:
https://github.com/mfem/mfem/blob/bf62d2923cc7390c0c78035dcebf3a6789187264/fem/gridfunc.cpp#L761-L770


For other DG integrators the face-to-volume transformation is unnecessarily repeated in the call to `Trans.SetIntPoint(&ip);`, e.g. here: https://github.com/mfem/mfem/blob/bf62d2923cc7390c0c78035dcebf3a6789187264/fem/bilininteg.cpp#L2573-L2582

The integration points for `Trans.Elem1` and `Trans.Elem2` (that are needed for coefficient evaluation later) can be set explicitly to `eip1` and `eip2`. Alternatively, we can do what `Trans.SetIntPoint(&ip);` does now and reuse the volume points stored internally in the `FaceElementTransformations` to call `CalcShape`, however, to make it clear that all (face and element) points are set, I suggest to rename the method to something that distinguishes it from the base class version of `SetIntPoint`, e.g. call it `SetAllIntPoints` or something like that which is more expressive.

This renamed method should then replace calls to the current `FaceElementTransformations::SetIntPoint` but only when appropriate, e.g. here:
https://github.com/mfem/mfem/blob/bf62d2923cc7390c0c78035dcebf3a6789187264/fem/gridfunc.cpp#L755

which will make it clear that we are setting all quadrature points, not just the face one. This is one place where I was confused by the name and assumed that we're setting just setting the face point.

psocratis commented:
@mlstowell, @v-dobrev, thank you both for the detailed explanations...I need some time to think this a bit more.

v-dobrev commented:
> @mlstowell, @v-dobrev, thank you both for the detailed explanations...I need some time to think this a bit more.

Maybe this will help: I think the main disagreement is about renaming the method `FaceElementTransformations::SetIntPoint` introduced recently in another PR by @mlstowell. I want to change it and @mlstowell does not. 😄

The rest of the concerns I mentioned above can be resolved independently of that, and I think we can find (and agree on) a solution for those once the renaming question is settled.

psocratis commented:
So...after thinking about it a bit more...and discussing with both of you...I think I prefer renaming the method. It seems more clear, instead of having the same method doing different things depending on the case. 

I guess there is always the risk of a developer of running into nasty bugs, but if each method is well and clearly documented, the risk should be minimal.

v-dobrev commented:
@mlstowell, @psocratis, @vladotomov, I thought of a possible compromise for the disagreement between @mlstowell and myself.

Here it is: we keep `FaceElementTransformations::SetIntPoint` as it is now (sets all, face+elem1+elem2, quadrature points) and we add a new method, `FaceElementTransformations::SetAllIntPoints`, that is just a more expressive name for the same operation -- it's an inline method that just calls `SetIntPoint`. With that, everywhere in MFEM we just use the more expressive name, `SetAllIntPoints`, for better clarity. On the other hand, if a user for some reason decides to call `FaceElementTransformations::SetIntPoint`, they'll still get all (face+elem) points set.

What do you think?

vladotomov commented:
This is a tricky one.

For me the problem is that we have a class, `FaceElementTransformations`, containing two detached pointers which are ultimately `Mesh::Transformation` and `Mesh::Transformation2`. In this situation it's difficult to predict all possible cases.

Mark's concern is of course valid. Calling `ElementTransformation::SetIntPoint` makes the three transformations point to different quadrature points, possibly leading to bugs. But on the other hand, the current implementation in `FaceElementTransformations::SetIntPoint` could lead to exactly the same problem -- if one is using `Mesh::Transformation` for something else, all of a sudden his `ElementTransformation` is using a different quadrature point. So even if `SetIntPoint` was `virtual`, I wouldn't be sure what's its best implementation in `FaceElementTransformations`.

I'd do a method with a different name, and no function `FaceElementTransformations::SetIntPoint` (this method can also lead to bugs when used with a pointer to the base class).

mlstowell commented:
I think this is a reasonable compromise.  Thank you @v-dobrev .

I'll need to give some thought to @vladotomov 's concern.  Reusing the `ElementTransformation` at different integration points was actually the initial problem which lead to this solution.  I don't think we've exposed ourselves to this problem again but it's worth double-checking.

mlstowell commented:
I've just put in the `FaceElementTransformations::SetAllIntPoints` member function and I've switched all but one of the relevant integrators and `GridFunction::GetValue` related methods to use this more expressive member function.  I also added accessor methods so we could use the `IntegrationPoint` objects stored inside the `FaceElementTransformations` objects as @v-dobrev suggested.  I think this removes any redundant point transformation operations and makes the integrators function with wider variety of `GridFunction` derived coefficients.

The one integrator that I didn't change is the `NormalTraceJumpIntegrator`.  This integrator calls `Trans.Loc1.Transf.SetIntPoint(&ip);`.  It is the only place I found where this `SetIntPoint` method is called on the `ElementTransformation` object stored inside an `IntegrationPointTransformation` object.  I'm not entirely sure why this is being called so I thought it was best to leave it alone.

I think this is ready to go pending any reviewer feedback.  Thanks @v-dobrev , @vladotomov , and @psocratis for your help so far.

v-dobrev commented:
LGTM

v-dobrev commented:
Re-merged in `next`.

v-dobrev commented:
@mlstowell, I added a few small changes -- take a look and see if they look good to you.

@psocratis, @vladotomov, how does this last version look to you?

@vladotomov, if I understand you, the concern you expressed above is that the internal members like `Mesh::Transformation` are used by multiple `Mesh` methods and it is not clear that, for example, the object returned by `Mesh::GetElementTransformation(int)` is then made invalid by a call to `Mesh::GetFaceElementTransformations`. I don't see a robust solution (that allows us to catch potential conflicts like these) that does not involve some interface change. Probably, the best we can do short term is to add more explicit documentation for this. Something like: "Warning: The object returned by this call invalidates the objects returned by previous calls to the following methods ...; and vice-versa, a later call to any of these methods will invalidate the object returned by this method."

mlstowell commented:
The changes all look good to me, @v-dobrev .  Thanks!

psocratis commented:
Changes look good to me too!

vladotomov commented on fem/eltrans.hpp:
```diff
@@ -466,23 +496,56 @@ class FaceElementTransformations : public IsoparametricTransformation
+       in this special case, does the same thing. This function can be used for
+       greater code clarity.
+   */
+   inline void SetAllIntPoints(const IntegrationPoint *face_ip)
```
I find it confusing to have functions with very different names `SetIntPoint` and `SetAllIntPoints` that actually do the same thing. If used with a `FaceElementTransformation` object, they both do the same. If used with a `ElementTransformation` pointer, they both don't exist. Anyway, it's not a big deal.

v-dobrev commented:
Note: this PR was merged into `master` indirectly: it was first merged it into #1529 which was then merged into `master`.
